In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

from pandas.plotting import scatter_matrix
from sklearn.decomposition import PCA # PCA

In [2]:
# 데이터 불러오기
data = pd.read_csv("project_data.csv")

## 첫번째 분석 방법

### 1) 이상치 제거
### 2) 정규화
### 3) PCA
### 4) 로지스틱을 이용한 분류

In [3]:
data1 = data

In [4]:
# value - mean의 절대값이 5시그마 밖의 값이면 이상치라 가정하고 제거
for i in data1.iloc[:,1:30]:
    data1 = data1[~(abs(data1[i] - data1[i].mean()) > 5*data1[i].std())] 

In [5]:
# X, Y로 나누고 X값들을 정규화
X = data1.iloc[:,1:30]
Y = data1.iloc[:,30]
for i in X:
    X[i] = (X[i] - X[i].mean())/X[i].std()

In [6]:
# PCA
pca_num = PCA(n_components=29)
X_PCA_T = pca_num.fit_transform(X)
X_PCA = pca_num.fit(X)

X_PCA_sum = np.cumsum(np.round(X_PCA.explained_variance_ratio_, decimals=8)*100) 
X_PCA_sum_varc = np.array(np.transpose(X_PCA_sum[:,np.newaxis])) 

X_PCA_sum_varc

array([[ 7.623337, 14.352209, 20.633861, 26.349525, 31.617967, 36.534743,
        41.3223  , 45.939378, 50.463384, 54.809141, 59.039765, 63.003636,
        66.77815 , 70.394455, 73.951452, 77.480637, 80.995487, 84.416027,
        87.800113, 90.622305, 92.593778, 94.37697 , 95.80591 , 97.172102,
        98.294107, 99.189599, 99.956628, 99.997534, 99.999998]])

위 결과에서 17번까지의 특징을 이용하면 전체 데이터의 80%를 설명할 수 있다

In [7]:
pc_y = np.c_[X_PCA_T,Y]
# 데이터프레임으로 변경
Data = pd.DataFrame(pc_y)

X = Data.iloc[:,0:17]
Y = Data.iloc[:,29]

In [8]:
from sklearn.model_selection  import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report

X_train,X_test,y_train,y_test=train_test_split(X,Y,test_size=0.5,random_state=1)

Log = LogisticRegression().fit(X_train, y_train)

Log_target_predict = Log.predict(X_test)

print(classification_report(y_test, Log_target_predict))

C:\Users\sunni\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


              precision    recall  f1-score   support

         0.0       1.00      1.00      1.00     93351
         1.0       0.00      0.00      0.00        28

   micro avg       1.00      1.00      1.00     93379
   macro avg       0.50      0.50      0.50     93379
weighted avg       1.00      1.00      1.00     93379



C:\Users\sunni\Anaconda3\lib\site-packages\sklearn\metrics\classification.py:1143: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)


위 결과를 봤을 때 어떤 데이터를 넣어도 Class = 0으로 예측을 한 것을 알 수 있다.

결과적으로 Class가 1인 데이터를 하나도 분류하지 못했다.

- 발견한 것 : 이상치 제거할 때 Class 값이 1인 데이터들이 많이 사라진 것을 발견

## 두번째 분석 방법

### 1) 정규화
### 2) PCA
### 3) 로지스틱을 이용한 분류

In [9]:
data2 = data

In [10]:
# X, Y로 나누고 X값들을 정규화
X = data2.iloc[:,1:30]
Y = data2.iloc[:,30]
for i in X:
    X[i] = (X[i] - X[i].mean())/X[i].std()

In [11]:
# PCA
pca_num = PCA(n_components=29)
X_PCA_T = pca_num.fit_transform(X)
X_PCA = pca_num.fit(X)

X_PCA_sum = np.cumsum(np.round(X_PCA.explained_variance_ratio_, decimals=8)*100) 
X_PCA_sum_varc = np.array(np.transpose(X_PCA_sum[:,np.newaxis])) 

X_PCA_sum_varc

array([[  6.821957,  10.472614,  14.054332,  17.581883,  21.059338,
         24.534958,  28.005411,  31.473655,  34.935393,  38.39431 ,
         41.84903 ,  45.301213,  48.746796,  52.188403,  55.628551,
         59.061438,  62.492848,  65.919956,  69.344699,  72.766799,
         76.186152,  79.598404,  83.005469,  86.410194,  89.806455,
         93.191683,  96.556926,  99.865013, 100.      ]])

위 결과에서 23번까지의 특징을 이용하면 전체 데이터의 80%를 설명할 수 있다

In [12]:
pc_y = np.c_[X_PCA_T,Y]
# 데이터프레임으로 변경
Data = pd.DataFrame(pc_y)

X = Data.iloc[:,0:23]
Y = Data.iloc[:,29]

In [13]:
from sklearn.model_selection  import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report

X_train,X_test,y_train,y_test=train_test_split(X,Y,test_size=0.5,random_state=1)

Log = LogisticRegression().fit(X_train, y_train)

y_pred = Log.predict(X_test)

print(classification_report(y_test, y_pred))

C:\Users\sunni\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


              precision    recall  f1-score   support

         0.0       1.00      1.00      1.00     99501
         1.0       0.79      0.58      0.67       181

   micro avg       1.00      1.00      1.00     99682
   macro avg       0.89      0.79      0.83     99682
weighted avg       1.00      1.00      1.00     99682



In [14]:
from sklearn import metrics
cnf_matrix = metrics.confusion_matrix(y_test, y_pred)
cnf_matrix

array([[99473,    28],
       [   76,   105]], dtype=int64)

위 결과를 봤을 때 Class = 1인 데이터를 받았을때 1이라고 할 확률이 58%이므로 좋지도 나쁘지도 않은 결과이다

## 세번째 분석 방법

### 1) 정규화
### 2) 로지스틱을 이용한 분류

In [15]:
data3 = data

In [16]:
# X, Y로 나누고 X값들을 정규화
X = data3.iloc[:,1:30]
Y = data3.iloc[:,30]
for i in X:
    X[i] = (X[i] - X[i].mean())/X[i].std()

In [17]:
from sklearn.model_selection  import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report

X_train,X_test,y_train,y_test=train_test_split(X,Y,test_size=0.5,random_state=1)

Log = LogisticRegression().fit(X_train, y_train)

y_pred = Log.predict(X_test)

print(classification_report(y_test, y_pred))

C:\Users\sunni\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


              precision    recall  f1-score   support

           0       1.00      1.00      1.00     99501
           1       0.78      0.57      0.66       181

   micro avg       1.00      1.00      1.00     99682
   macro avg       0.89      0.78      0.83     99682
weighted avg       1.00      1.00      1.00     99682



In [18]:
from sklearn import metrics
cnf_matrix = metrics.confusion_matrix(y_test, y_pred)
cnf_matrix

array([[99472,    29],
       [   78,   103]], dtype=int64)

위 결과를 봤을 때 Class = 1인 데이터를 받았을때 1이라고 할 확률이 57%이므로 PCA를 이용한 결과와 거의 동일하다고 볼 수 있다.

## 네번째 분석 방법

### 1) 정규화
### 2) 분산이 높은 특징 추출(직접)
### 3) 로지스틱을 이용한 분류

In [19]:
data4 = data

In [20]:
# X, Y로 나누고 X값들을 정규화
X = data4.iloc[:,1:30]
Y = data4.iloc[:,30]
for i in X:
    X[i] = (X[i] - X[i].mean())/X[i].std()

In [21]:
import statsmodels.api as sm
x2 = sm.add_constant(X)
model = sm.OLS(Y,x2)
result = model.fit()
print(result.summary())

                            OLS Regression Results                            
Dep. Variable:                  Class   R-squared:                       0.511
Model:                            OLS   Adj. R-squared:                  0.511
Method:                 Least Squares   F-statistic:                     7193.
Date:                Wed, 22 May 2019   Prob (F-statistic):               0.00
Time:                        17:07:48   Log-Likelihood:             4.1946e+05
No. Observations:              199364   AIC:                        -8.389e+05
Df Residuals:                  199334   BIC:                        -8.385e+05
Df Model:                          29                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.0018   6.61e-05     27.013      0.0

위 결과에서 각 특징별 t값이 있는데 t값이 높다는 것은 그 특징이 유의미한 존재임을 말해준다.

먼저 각 특징별 t값의 절댓값이 100이 넘는 특징을 선택해 분류해 보려고 한다.

V10,V12,V14,V16,V17 다섯개의 특징만 사용

In [22]:
X = data.iloc[:,[10,12,14,16,17]]
Y = data.iloc[:,30]

In [23]:
from sklearn.model_selection  import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report

X_train,X_test,y_train,y_test=train_test_split(X,Y,test_size=0.5,random_state=1)

Log = LogisticRegression().fit(X_train, y_train)

y_pred = Log.predict(X_test)

print(classification_report(y_test, y_pred))

C:\Users\sunni\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


              precision    recall  f1-score   support

           0       1.00      1.00      1.00     99501
           1       0.88      0.56      0.69       181

   micro avg       1.00      1.00      1.00     99682
   macro avg       0.94      0.78      0.84     99682
weighted avg       1.00      1.00      1.00     99682



In [24]:
from sklearn import metrics
cnf_matrix = metrics.confusion_matrix(y_test, y_pred)
cnf_matrix

array([[99487,    14],
       [   79,   102]], dtype=int64)

여전히 Class = 1인 데이터를 받았을때 1이라고 할 확률이 56%이므로 크게 달라진 점은 없다.

## 다섯번째 분석 방법

### 1) 정규화
### 2) 분산이 높은 특징 추출(컴퓨터)
### 3) 로지스틱을 이용한 분류

In [25]:
data5 = data

In [26]:
# X, Y로 나누고 X값들을 정규화
X = data5.iloc[:,1:30]
Y = data5.iloc[:,30]
for i in X:
    X[i] = (X[i] - X[i].mean())/X[i].std()

In [27]:
from sklearn.model_selection  import train_test_split
from sklearn.linear_model import LinearRegression

X_train,X_test,y_train,y_test=train_test_split(X,Y,test_size=0.5,random_state=1)
Log = LinearRegression().fit(X_train, y_train)

In [28]:
from sklearn.svm import SVR 
from sklearn.feature_selection import RFE 
from sklearn import datasets
selector = RFE(Log, 5).fit(X_train, y_train) #  전체 특징 중 중요한 5개의 특징을 정해줌
print(selector.support_) 
print(selector.ranking_)

[False False False False False False False False False  True False  True
 False  True False  True  True False False False False False False False
 False False False False False]
[ 8  9  2  5 10 12  3 15  7  1  4  1 18  1 20  1  1  6 14 25 13 24 23 17
 22 21 16 19 11]


In [29]:
X = data.iloc[:,[9,11,13,15,16]]
Y = data.iloc[:,30]

In [30]:
from sklearn.model_selection  import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report

X_train,X_test,y_train,y_test=train_test_split(X,Y,test_size=0.5,random_state=1)

Log = LogisticRegression().fit(X_train, y_train)

y_pred = Log.predict(X_test)

print(classification_report(y_test, y_pred))

C:\Users\sunni\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


              precision    recall  f1-score   support

           0       1.00      1.00      1.00     99501
           1       0.91      0.49      0.63       181

   micro avg       1.00      1.00      1.00     99682
   macro avg       0.95      0.74      0.82     99682
weighted avg       1.00      1.00      1.00     99682



In [31]:
from sklearn import metrics
cnf_matrix = metrics.confusion_matrix(y_test, y_pred)
cnf_matrix

array([[99492,     9],
       [   93,    88]], dtype=int64)

Class = 1인 데이터를 받았을때 1이라고 할 확률이 49%이므로 분석 결과가 더 낮아졌다.

## 여섯번째 분석 방법

### 1) 정규화
### 2) 직관적으로 분석

In [32]:
X = data

In [33]:
y_test = X.iloc[:,30]
y_pred = np.zeros((199364,))

현재 문제는 Class 값이 1과 0을 구분하는 것이기 때문에 Class = 1과 0을 그룹화 한 뒤 각 특징별 분산의 차이를 확인했는데 V7,V8,V17의 분산이 차이가 많이 나는 것을 발견했습니다.

그래서 위 세개의 특징 값으로 각 데이터 값을 비교해 봤더니 육안으로 차이가 존재하는 것을 확인했습니다.

결과적으로 각 특징별 데이터 값을 가지고 제가 임의로 지정한 값 이상일 경우에는 모두 Class가 1로 간주했습니다.

아래는 위에 말한 내용을 함수로 정리한 내용입니다.

In [34]:
def my_Analysis_method(data):
    b1 = set(data[abs(data.iloc[:,17]) > 1.5].index)
    b2 = set(data[abs(data.iloc[:,7]) > 6].index)
    b3 = set(data[abs(data.iloc[:,8]) > 6].index)

    return b1 | b2 | b3

In [35]:
result = my_Analysis_method(X)

In [36]:
for i in result:
    y_pred[i] = 1

In [37]:
from sklearn import metrics
cnf_matrix = metrics.confusion_matrix(y_test, y_pred)
cnf_matrix

array([[190269,   8739],
       [    60,    296]], dtype=int64)

In [38]:
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      0.96      0.98    199008
           1       0.03      0.83      0.06       356

   micro avg       0.96      0.96      0.96    199364
   macro avg       0.52      0.89      0.52    199364
weighted avg       1.00      0.96      0.98    199364



결과적으로 실제 값이 Class가 1일 때 1로 예측할 확률이 83%로 증가하게 되었습니다.

물론 위의 여러 결과물 보다 Accuracy가 떨어지지만 Class가 1인 데이터를 찾는것에 focus를 둔다면 위의 분석 방법들 보다 이상적인 분석방법이라 할 수 있다고 생각합니다.